ra/ra.go

```diff
@@ -1424,18 +1392,54 @@ func (ra *RegistrationAuthorityImpl) NewOrder(ctx context.Context, req *rapb.New
 		return nil, err
 	}
 
-	// TODO(#2955): Replace this with the batched methods
+	now := ra.clk.Now().UnixNano()
+	existingAuthz, err := ra.SA.GetAuthorizations(ctx, &sapb.GetAuthorizationsRequest{
+		RegistrationID: req.RegistrationID,
+		Now:            &now,
+		Domains:        parsedCSR.DNSNames,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	gotAuthzFor := make(map[string]bool, len(parsedCSR.DNSNames))
+	for _, v := range existingAuthz.Authz {
+		gotAuthzFor[*v.Domain] = true
+		order.Authorizations = append(order.Authorizations, *v.Authz.Id)
+	}
+
+	var newAuthz []*corepb.Authorization
 	for _, name := range parsedCSR.DNSNames {
-		authz, err := ra.NewAuthorization(ctx, core.Authorization{
-			Identifier: core.AcmeIdentifier{
-				Type:  core.IdentifierDNS,
-				Value: name,
-			},
-		}, *req.RegistrationID)
+		if _, present := gotAuthzFor[name]; present {
+			continue
+		}
+		identifier := core.AcmeIdentifier{Value: name, Type: core.IdentifierDNS}
+		if err := ra.PA.WillingToIssue(identifier); err != nil {
```

Comment:
We should probably check WillingToIssue before calling `GetValidAuthorizations`, since it's cheap, and can change after creating valid authorizations.